### DISCUSSION TITLE
Make Localizer more convenient to inheritance

### DISCUSSION THREAD

**Discussion by a-sanders** [perpetualKid/GetText.NET#29] [Ideas]
Hello @perpetualKid,
Is it possible to ask you to make higher visibility (private --> protected) of some Localizer's methods and properties? I'd like to inherit Localizer class (GetText.WindowsForms) to be able to expand some functionality.

This is the Localizer.cs file with desired changes: 
[Localizer.cs](https://github.com/perpetualKid/GetText.NET/files/6061993/Localizer.cs.zip)

Thank you.
Regards.

**Comment by perpetualKid** [comment]
I could see this working both ways, opening up to enable inheritance, or (depending on your changes to functionality) incorporating this in the Localizer itself.

**Reply by a-sanders** [reply]
Sounds good!
So could you please make the changes that I attached in my previous message?

**Reply by perpetualKid** [reply]
looking at your code, trying to understand what you gain here? You are basically plugging into the recursive iteration of child controls, is that what you want? Wouldn't help i.e. the case of the ContextMenu.
Is the idea to skip calling HandleControl based on certain condition?

Also, why would you want ccess to  `protected readonly Control root;`?

**Reply by a-sanders** [reply]
@perpetualKid this is exactly your code just a bit changed to be able to create inheritance classes. Your logic was not changed at all. Namely, I've changed visibility private --> protected for a couple of things and extracted a part of IterateControls to the new method HandleControl

```diff
18c18
<         private readonly Control root;
---
>         protected readonly Control root;
111c111
<         private enum IterateMode
---
>         protected enum IterateMode
149c149
<         private void IterateControls(Control control, IterateMode mode)
---
>         protected virtual void IterateControls(Control control, IterateMode mode)
155a156,160
>             HandleControl(control, mode);
>         }
> 
>         protected void HandleControl(Control control, IterateMode mode)
>         {
```

These changes will allow me to inherit and handle exclude-filter for controls.

```C#
        ....
        
        public List<Type> ExcludeControlClasses = new List<Type> {
            typeof(TextBoxBase),
            typeof(UpDownBase)
        };
        
        ....
        
        protected override void IterateControls(Control control, IterateMode mode)
        {
            foreach (Control child in control.Controls)
            {
                IterateControls(child, mode);
            }

            if ( !ExcludeControlClasses.Exists(x => control.GetType().IsSubclassOf(x)) )
                HandleControl(control, mode);
        }
```

**Reply by a-sanders** [reply]
> Also, why would you want ccess to `protected readonly Control root;`?

I did not want to. But it was required to get the protected IterateControls method...

**Reply by a-sanders** [reply]
> Wouldn't help i.e. the case of the ContextMenu

The ContextMenu can be accessible from the Components list only
```C#
foreach (Component component in control.Components)
{
    ...
}
```

**Reply by perpetualKid** [reply]
I've published the changes here https://github.com/perpetualKid/GetText.NET/tree/LocalizerInheritance

can you check if this covers your requirements, than I'll publish a package.

**Reply by a-sanders** [reply]
Thank you. It's exactly what I meant.
BTW let's go ahead and make the other methods protected, namely InitFromContainer, IterateControlHandler, the rest IterateControls, and the IterateToolStripItems? It would be convenient for further expanding the functionality.

**Reply by perpetualKid** [reply]
I've updated the other methods as well and also make them virtual so you may override them. 

However IterateControlHandler is kept private, otherwise would need to make LocalizableObjectAdapter as well. I'd rather keep this for a redesign/re-engineering, I'm not really happy with this current approach.

I'll publish a package next days, lmk if there is anything else which needs to be covered.

**Reply by a-sanders** [reply]
Thank you, it's almost all that it needed to be able flexible expand the Localizer.

I found a way to implement localizing ContextMenuStrip but I still need s few changes in Localizer.cs:

```diff
98c98
<         public void Localize()
---
>         public virtual void Localize()
103c103
<         public void Revert()
---
>         public virtual void Revert()
117c117
<         private void IterateControlHandler(LocalizableObjectAdapter adapter, IterateMode mode)
---
>         protected void IterateControlHandler(LocalizableObjectAdapter adapter, IterateMode mode)
``` 

And yes - it will require to make LocalizableObjectAdapter.cs as well:

```diff
12c12
<     internal class LocalizableObjectAdapter
---
>     public class LocalizableObjectAdapter
28c28
<         public void Localize(ICatalog catalog)
---
>         public virtual void Localize(ICatalog catalog)
50c50
<         public void Revert()
---
>         public virtual void Revert()
85c85
<         private void LocalizeProperty(ICatalog catalog, string propertyName)
---
>         protected virtual void LocalizeProperty(ICatalog catalog, string propertyName)
92c92
<         private void RevertProperty(string propertyName)
---
>         protected virtual void RevertProperty(string propertyName)
```

**Reply by perpetualKid** [reply]
As I mentioned [before](https://github.com/perpetualKid/GetText.NET/issues/27#issuecomment-787987767), you can localize the ContextMenu already by passing as control to another call of Localize
`Localizer.Localize(contextMenuStrip, catalog, store);` 

Else it may be easier to implement your own WindForms Localizer instead of deriving, if you want to modify that much. Just use this code as base, it's free&open by MIT license

**Reply by a-sanders** [reply]
> Localizer.Localize(contextMenuStrip, catalog, store);

I use this way from the beginning but this is the worst way in my case - I have 13 context menus on only one form. 
The preferred way is to use the class Localizer with basic functionality and good inheritance ability. Isn't that what object-oriented design is all about? :innocent:

In any case, I would like to express my gratitude to you for what you are doing! :thumbsup:

**Reply by perpetualKid** [reply]
I've updated the code, and also replaced the IterateMode enum with a simple bool. Have a look, and when ok, I'll publish a new package version

**Reply by a-sanders** [reply]
Everything is fine
